File: BST/BST.c
/*
   Binary Search  Tree in C Programming
   
   Containts : InsertNode Function
               Inorder Function
               Preorder Function
               Postorder Function
               
*/

#include<stdio.h>
#include<stdlib.h>

struct Node
{
  struct Node * lchild;	
  int Data;
  struct Node * rchild;
};

typedef struct Node NODE;
typedef struct Node * PNODE;
typedef struct Node ** PPNODE;

void InsertNode(PPNODE Root , int iNo)
{
   PNODE Newn = (PNODE)malloc(sizeof(NODE));

   Newn->lchild= NULL;
   Newn->rchild = NULL;
   Newn->Data = iNo; 

   if((*Root) == NULL)
   {
      (*Root) = Newn;
   }
   else
   {
      PNODE Temp = *Root;	   
     
      while(1)
      {
         if(iNo > Temp->Data)
	 {
            if(Temp->rchild == NULL)
	    {
               Temp->rchild = Newn;
	       break;
	    }

	    Temp = Temp->rchild;
	 }
         else if(iNo < Temp->Data)
	 {
            if(Temp->lchild == NULL)
	    {
               Temp->lchild = Newn;
	       break;
	    }

	    Temp = Temp->lchild;
	 }
	 else if(iNo == Temp->Data)
	 {
            printf("No Duplicate Elements allowed in BST\n");
	    free(Newn);
	    break;
	 }
      }
   }
}

void Inorder(PNODE Root)
{
   if(Root != NULL)
   {
      Inorder(Root->lchild);
      printf("%d\t",Root->Data);
      Inorder(Root->rchild);
   }

}

void Preorder(PNODE Root)
{
   if(Root != NULL)
   {
      printf("%d\t",Root->Data);
      Preorder(Root->lchild);
      Preorder(Root->rchild);
   }
 
}

void Postorder(PNODE Root)
{
  if(Root != NULL)
  {
     Postorder(Root->lchild);
     Postorder(Root->rchild);
     printf("%d\t",Root->Data);
  }
}

int main()
{
  PNODE Head = NULL;

  
  InsertNode(&Head , 50);
  InsertNode(&Head , 30);
  InsertNode(&Head , 20);
  InsertNode(&Head , 40);
  InsertNode(&Head , 70);
  InsertNode(&Head , 60);
  InsertNode(&Head , 80);



  Inorder(Head);
  printf("\n");

  Preorder(Head);
  printf("\n");

  Postorder(Head);
  printf("\n");

  return 0;
}

File: BST/BST.java
import java.util.*;


class Node
{
   Node Left;
   Node Right;
   
   int Data;
   
   public Node(int No)
   {
      this.Data = No;
   }   
}


class BST
{
   public static void main(String arg[])
   {
       Scanner sobj = new Scanner(System.in);
       
       Node Root = CreateTree();
       InOrder(Root);
       System.out.println();
       PreOrder(Root);
       System.out.println();
       PostOrder(Root);
       System.out.println();
   }
   
   static Node CreateTree()
   {
   
     Scanner sobj = new Scanner(System.in);
      
      Node Root = null;
      
       System.out.println("Enter Data : ");
       int Data = sobj.nextInt();
       
       if(Data == -1)
       {
          return null;
       }
       
       Root = new Node(Data);
       
       System.out.println("Enter Left for "+Data);
       Root.Left = CreateTree();
       
       System.out.println("Enter right for "+Data);
       Root.Right = CreateTree();
       
       return Root;
       
   }
   
   static void InOrder(Node Root)
   {
      if(Root == null)
      {
         return;
      }
      
      InOrder(Root.Left);
      System.out.print(" "+Root.Data);
      InOrder(Root.Right);
   }
   
   static void PreOrder(Node Root)
   {
      if(Root == null)
      {
         return;
      }
      
      System.out.print(" "+Root.Data);
      InOrder(Root.Left);
      InOrder(Root.Right);
   }
   
   static void PostOrder(Node Root)
   {
      if(Root == null)
      {
         return;
      }
      
      
      InOrder(Root.Left);
      InOrder(Root.Right);
      System.out.print(" "+Root.Data);
   }
   
}
